synapse/handlers/sync.py

```diff
@@ -1010,6 +1039,13 @@ async def compute_state_delta(
                             ),
                         )
 
+            # At this point, `state_ids` includes the memberships of all event senders
+            # in the timeline when `lazy_load_members` is enabled. This is because we
+            # may not have sent the memberships in a previous sync.
```

Comment:
```suggestion
            # At this point, if `lazy_load_members` is enabled, `state_ids` includes
            # the memberships of all event senders in the timeline. This is because we
            # may not have sent the memberships in a previous sync.
```

I think?

Reply:
Is this true for all three sync types?

Reply:
I believe so?

Assuming the docstring for the return value is right, `state_ids` (which is almost the return value) has to have that property after all three branches, otherwise lazy loading would be broken (this reasoning is a little backwards).

---

It's a lot more confusing to try to reason forwards:

For an initial sync, `state_ids` must include the memberships of event senders, otherwise it's not an initial sync?
Less handwavey:
`current_state_ids` contains the memberships of event senders at the end of the timeline, and `state_ids` initially contains the memberships of event senders as seen at the start of the timeline. The branch for empty timelines can be ignored because there are no event senders in that case.
`_calculate_state` returns `(timeline end state | timeline start state) - (previous sync state - memberships in timeline start) - state in timeline`. The previous sync state is `{}`, so `_calculate_state` returns the memberships of event senders from the state at the start of the timeline.

For a gappy sync:
`state_at_timeline_start` contains the memberships of event senders at the start of the timeline
`state_at_previous_sync` contains all memberships at the end of the previous sync
`current_state_ids` contains all memberships at the end of the timeline.
`_calculate_state` returns `(timeline end state | timeline start state) - (previous sync state - memberships in timeline start) - state in timeline`.
The `previous sync state - memberships in timeline start` term excludes the memberships of event senders.
(this branch is a mess)

For an incremental sync, we fetch the memberships of event senders explicitly.

Reply:
I'm somewhat tempted to rename some of these variables...

Reply:
> It's a lot more confusing to try to reason forwards:

ok, yes, that sounds right - thank you for ploughing through that!